### PR TITLE
Feature/idp integration

### PR DESCRIPTION
**Pull request by olye0216**
idp integration

### CODE REVIEW

File: src/main/java/com/icthh/xm/gate/gateway/BeariringModeFilter.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import com.icthh.xm.gate.security.oauth2.IdpConfigRepository;
+import com.netflix.zuul.ZuulFilter;
+import com.netflix.zuul.context.RequestContext;
+import javax.servlet.http.HttpServletRequest;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class BeariringModeFilter extends ZuulFilter {
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/gate/gateway/BeariringModeFilter.java:20]: @olye0216 as I remember we have renamed the bearing feature. Please use the new name. 
But anyway seems this filter checks only stateful feature. So please choose correct name for the class.

Comment (olye0216) [reply src/main/java/com/icthh/xm/gate/gateway/BeariringModeFilter.java:20]: Renamed to IdpStatefulModeFilter

File: src/test/java/com/icthh/xm/gate/web/rest/LogsResourceIntTest.java [hunk 2]
```diff
@@ -48,7 +48,7 @@ public void setup() {
     public void getAllLogs()throws Exception {
         restLogsMockMvc.perform(get("/management/logs"))
             .andExpect(status().isOk())
-            .andExpect(content().contentType(MediaType.APPLICATION_JSON_UTF8_VALUE));
+            .andExpect(content().contentType(MediaType.APPLICATION_JSON_VALUE));
```
Comment (vkirichenko) [inline src/test/java/com/icthh/xm/gate/web/rest/LogsResourceIntTest.java:51]: why this media type was changed?

Comment (olye0216) [reply src/test/java/com/icthh/xm/gate/web/rest/LogsResourceIntTest.java:51]: Media type MediaType.APPLICATION_JSON_UTF8_VALUE marked as deprecated in sources.

File: src/main/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandler.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+                                           JHipsterProperties jhipsterProperties) {
+        this.objectMapper = objectMapper;
+        this.tenantContextHolder = tenantContextHolder;
+        this.restTemplate = restTemplate;
+        this.idpConfigRepository = idpConfigRepository;
+        this.jhipsterProperties = jhipsterProperties;
+    }
+
+    @Override
+    public void onAuthenticationSuccess(HttpServletRequest request,
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandler.java:72]: can we cover this method by lep? (if it possible)

File: src/main/java/com/icthh/xm/gate/security/oauth2/IdpClientRepository.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    private static Map<String, ClientRegistration> mapRegistrationIdToClientRegistration(
+        Collection<ClientRegistration> registrations) {
+        return registrations
+            .stream()
+            .peek(registration -> Assert.notNull(registration, "registrations cannot contain null values"))
+            .collect(Collectors.toUnmodifiableMap(ClientRegistration::getRegistrationId, Function.identity()));
+    }
+
+    @Override
+    public ClientRegistration findByRegistrationId(String registrationId) {
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/gate/security/oauth2/IdpClientRepository.java:64]: pls add Lep annotation if it possible here

File: src/main/java/com/icthh/xm/gate/gateway/IdpStatefulModeFilter.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @Override
+    public String filterType() {
+        return "pre";
+    }
+
+    @Override
+    public int filterOrder() {
+        return 10000;
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/gate/gateway/IdpStatefulModeFilter.java:57]: magic number

File: src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+            .thenReturn(buildResponse(true));
+
+        this.idpAuthenticationSuccessHandler =
+            new IdpAuthenticationSuccessHandler(objectMapper, restTemplate, tenantContextHolder, idpConfigRepository, jhipsterProperties);
+
+        MockHttpServletResponse response = new MockHttpServletResponse();
+        idpAuthenticationSuccessHandler.onAuthenticationSuccess(null, response, buildAuthentication("email"));
+
+        //Capture and validate args
+        ArgumentCaptor<String> accessTokenUriCaptor = ArgumentCaptor.forClass(String.class);
```
Comment (sergeysenja1992) [inline src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java:125]: replace argument captor to eq or refEq validation

File: src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+        ArgumentCaptor<Object> uaaTokenRequestCaptor = ArgumentCaptor.forClass(Object.class);
+
+        verify(restTemplate, times(1)).exchange(
```
Comment (sergeysenja1992) [inline src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java:128]: times(1) not needed here

File: src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+            accessTokenUriCaptor.capture(),
```
Comment (sergeysenja1992) [inline src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java:129]: simple eq("http://uaa.com")

Comment (olye0216) [reply src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java:129]: Remark applied

File: src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java (hunk 7 above)
Comment (olye0216) [reply src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java:128]: Remark applied

File: src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java (hunk 6 above)
Comment (olye0216) [reply src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java:125]: Remark applied

File: src/main/java/com/icthh/xm/gate/security/oauth2/IdpConfigRepository.java [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+    public void onInit(String configKey, String configValue) {
+        updateIdpConfigs(configKey, configValue);
+    }
+
+    @Override
+    public void onRefresh(String updatedKey, String config) {
+        updateIdpConfigs(updatedKey, config);
+    }
+
+    private void updateIdpConfigs(String configKey, String config) {
```
[the remark this answers is not in the document]
Comment (olye0216) [reply src/main/java/com/icthh/xm/gate/security/oauth2/IdpConfigRepository.java:99]: Remark applied.

File: src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+        IdpPublicConfig idpPublicConfig = registerPublicConfigs(clientKeyPrefix, tenantKey, clientsAmount, true, true, false);
+        IdpPrivateConfig idpPrivateConfig = registerPrivateConfigs(clientKeyPrefix, tenantKey, clientsAmount, true, true);
+
+        validateRegistration(tenantKey, clientKeyPrefix, clientsAmount, idpPublicConfig, idpPrivateConfig);
+
+        TenantContextUtils.setTenant(tenantContextHolder, tenantKey);
+        jhipsterProperties.getSecurity().getClientAuthorization().setAccessTokenUri("http://uaa.com");
+
+        when(restTemplate.exchange(
+            anyString(),
```
Comment (sergeysenja1992) [inline src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java:112]: change any string to actual value

File: src/main/java/com/icthh/xm/gate/gateway/IdpStatefulModeFilter.java (hunk 5 above)
Comment (olye0216) [reply src/main/java/com/icthh/xm/gate/gateway/IdpStatefulModeFilter.java]: Remark applied.

File: src/main/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandler.java (hunk 3 above)
Comment (amedvedchuk) [reply src/main/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandler.java:72]: @sergeysenja1992 we decided not to add LEP here from a start as we do not see the cases of using it.
But in the future, we may get back to this question and add if need.

File: src/main/java/com/icthh/xm/gate/security/oauth2/IdpClientRepository.java (hunk 4 above)
Comment (amedvedchuk) [reply src/main/java/com/icthh/xm/gate/security/oauth2/IdpClientRepository.java:63]: @sergeysenja1992 we decided not to add LEP here from a start as we do not see the cases of using it.
But in the future, we may get back to this question and add if need.

File: src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java (hunk 10 above)
Comment (olye0216) [reply src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java:112]: Remark applied

File: src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+            eq(HttpMethod.POST),
+            any(),
```
[the remark this answers is not in the document]
Comment (olye0216) [reply src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java:114]: In this case, an object that stands behind "any()" is complex and validated in the code below, so it looks like overhead for this case to inline some object for inner equality validation.

[the remark this answers is not in the document]
Comment (sergeysenja1992) [reply src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java:114]: but body and content of this object criticality important for this function

**Review by vkirichenko — APPROVED** [review]

File: src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (vkirichenko) [reply src/test/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandlerUnitTest.java:114]: @sergeysenja1992 body is verified in separate block
`verify(restTemplate).exchange(
            eq("http://uaa.com"),
            eq(HttpMethod.POST),
            (HttpEntity<?>) uaaTokenRequestCaptor.capture(),
            (ParameterizedTypeReference<Map<String, Object>>) any()
        );`

**Review by amedvedchuk — APPROVED** [review]

**Pull request merged**